Sources/TuistCore/Graph/GraphTraverser.swift

```diff
@@ -603,6 +603,78 @@ public class GraphTraverser: GraphTraversing {
         )
     }
 
+    public func targetsWithExternalDependencies() -> Set<GraphTarget> {
+        Set(graph.dependencies.compactMap { fromDependency, toDependencies in
+            let fromNonExternalTarget: GraphTarget?
+            switch fromDependency {
+            case let .target(targetName, projectPath):
+                let project = graph.projects[projectPath]!
+                if project.isExternal {
+                    fromNonExternalTarget = nil
+                } else {
+                    let target = graph.targets[projectPath]![targetName]!
+                    fromNonExternalTarget = GraphTarget(path: projectPath, target: target, project: project)
+                }
+            case .bundle, .framework, .library, .packageProduct, .sdk, .xcframework:
+                fromNonExternalTarget = nil
+            }
+            guard fromNonExternalTarget != nil else { return nil }
+
+            let dependsOnExternalDependency = toDependencies.first { toDependency in
+                switch toDependency {
+                case let .target(_, projectPath):
+                    graph.projects[projectPath]!.isExternal
+                case .bundle, .framework, .library, .packageProduct, .sdk, .xcframework:
+                    false
+                }
+            } != nil
+
+            return dependsOnExternalDependency ? fromNonExternalTarget : nil
+        })
+    }
+
+    public func externalTargets() -> Set<GraphTarget> {
+        Set(graph.projects.compactMap { path, project in
+            project.isExternal ? (path, project) : nil
+        }.flatMap { projectPath, project in
+            let targets = graph.targets[projectPath, default: [:]].values
+            return targets.map { GraphTarget(path: projectPath, target: $0, project: project) }
+        })
+    }
+
+    public func orphanExternalDependencies() -> Set<GraphDependency> {
```

Comment:
This feels very specific to the pruning mapper – should this be defined as an extension there? I don't see how it could be useful in other places and it would be good if the `GraphTraverser` doesn't grow out of proportions.

Reply:
I'd leave it here, where we might introduce traversing optimizations in the future that this logic could benefit from.